src/data/row.rs

```diff
@@ -41,34 +41,40 @@ impl Row {
     }
 
     pub fn new(column_defs: &[ColumnDef], columns: &[String], values: &[Expr]) -> Result<Self> {
+        if !columns.is_empty() && values.len() != columns.len() {
+            return Err(RowError::DoNotMatchColumnsAndValues.into());
+        }
         if values.len() > column_defs.len() {
             return Err(RowError::TooManyValues.into());
         }
 
+        let columns: Box<dyn Iterator<Item = &String>> = if columns.is_empty() {
+            Box::new(column_defs.iter().map(|ColumnDef { name, .. }| name))
+        } else {
+            Box::new(columns.iter())
+        };
```

Comment:
`iter_enum::Iterator` would be quite a good choice to take it here.

```rust
#[derive(Iterator)]                                                    
enum Columns<I1, I2> {                                                 
    All(I1),                                                           
    Specified(I2),                                                     
}                                                                      
                                                                       
let columns = if columns.is_empty() {                                  
    Columns::All(column_defs.iter().map(|ColumnDef { name, .. }| name))
} else {                                                               
    Columns::Specified(columns.iter())                                 
};                                                                     
```

Reply:
A very nice change! thank you.